File: python/day1/day1_part2.py
import numpy as np

example_list1 = [3, 4, 2, 1, 3, 3]
example_list2 = [4, 3, 5, 3, 9, 3]

count_of_3 = example_list1.count(3)

sorted_list1 = np.sort(example_list1)
sorted_list2 = np.sort(example_list2)

deduped_example_list1 = list(dict.fromkeys(sorted_list1))
print(deduped_example_list1)

count_from_dict_thing = deduped_example_list1.count(3)


sorted_list2 = list(sorted_list2)

difference_scalar = 0
for i in deduped_example_list1:
    if i in sorted_list2:
        m = sorted_list2.count(i)
        difference_scalar += i*m
    else:
        difference_scalar += 0

print(f"Example scalar dif", difference_scalar)

# ---------------------------------------------------------------------------------------------- #

# Lists to store the numbers from each column
numbers_col1 = []
numbers_col2 = []

with open("python\day1\day1_input.txt", "r") as file:
    for line in file:
        # Split each line on whitespace and convert to integers
        num1, num2 = map(int, line.strip().split())
        numbers_col1.append(num1)
        numbers_col2.append(num2)

# Convert lists to numpy arrays
list1 = np.array(numbers_col1)
list2 = np.array(numbers_col2)

sorted_list1 = np.sort(list1)
sorted_list2 = np.sort(list2)

deduped_example_list1 = list(dict.fromkeys(sorted_list1))


sorted_list2 = list(sorted_list2)

difference_scalar = 0
for i in deduped_example_list1:
    if i in sorted_list2:
        m = sorted_list2.count(i)
        difference_scalar += i*m
    else:
        difference_scalar += 0

print(f"final difference for advent = ",difference_scalar)

File: rust/day1/day1_part2.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::collections::HashSet;

fn main() -> io::Result<()> {
    // Read the input file
    let path = Path::new("python/day1/day1_input.txt");
    let file = File::open(path)?;
    let reader = io::BufReader::new(file);

    // Store numbers from each column
    let mut numbers_col1: Vec<i32> = Vec::new();
    let mut numbers_col2: Vec<i32> = Vec::new();

    // Read and parse each line
    for line in reader.lines() {
        if let Ok(line) = line {
            let numbers: Vec<i32> = line
                .split_whitespace()
                .map(|s| s.parse().unwrap())
                .collect();
            
            if numbers.len() == 2 {
                numbers_col1.push(numbers[0]);
                numbers_col2.push(numbers[1]);
            }
        }
    }

    // Sort and deduplicate first list
    numbers_col1.sort();
    let deduped_list1: Vec<i32> = numbers_col1.iter()
        .cloned()
        .collect::<HashSet<i32>>()
        .into_iter()
        .collect::<Vec<i32>>();
    deduped_list1.sort();

    // Calculate scalar difference
    let mut difference_scalar = 0;
    for num in deduped_list1 {
        let count = numbers_col2.iter().filter(|&&x| x == num).count();
        difference_scalar += num * count as i32;
    }

    println!("Final difference for advent = {}", difference_scalar);
    Ok(())
} 